FLIGHT_DEVELOPER_GUIDE.md

```diff
@@ -0,0 +1,442 @@
+step in a flight is running and the server process is killed (e.g., the Kubernetes pod is
+deleted) the flight can be recovered. On recovery, the step that was running at the point
+of failure, is restarted.
+
+As a result, the `do()` and `undo()` methods of a step must be written to be
+idempotent; that is, they can be run multiple times and must have the same effect each
+time. Meeting that requirement often uses a different approach than writing a sequential program.
+
+### Database Transactions <a name="databasetransactions"></a>
+A step that consists of a database transaction is simple to write. The underlying database
+system provides the rollback of the transaction in the event of failure. However, it is
+still possible for a database transaction step to fail in the window between when the
+database transaction commits and when Stairway logs the completion of the step.
+
+You still have to make the step idempotent, so you have to handle that condition. Let's
+say you are inserting a new object into the database. This is best done by giving your
+target table a primary key column that is a UUID. Then you can run a sequence of steps
+something like this:
+ 1. Step that generates a random UUID and stores it in the working map; _OR_ hand a random
+ UUID into the flight as an input parameter. If this step is run more than once, it will
+ simply generate a different UUID.
+ 1. Step that runs a database transaction. The database transaction does:
+    1. Read the table to see if a row with the UUID already exists. If so, return
+    success. A previous run of the step inserted that row.
+    1. Insert the row into the table.
+    If this step is run more than once, either the database transaction will have rolled
+    back and the row will be re-inserted, or the database transaction will have committed
+    and the read will detect that the insert already happened successfully.
+
+
+A variation on this approach is for the step to trap the duplicate key exception from the
+database system and consider that a success.
+
+### Existence Checking <a name="existencechecking"></a>
+The database transaction case brings up the general problem of coordinating existence
+checking and idempotent steps. There are a few common pitfalls.
+
+#### Pitfall 1: deleting someone else's object
+Imagine that you want to create a thing named "foo". The thing might be a file, a cloud
+object, a database row, whatever. You write a step like this:
+
+```
+Step: Create a thing
+ do():
+   create "foo"
+ undo():
+   delete "foo"
+```
+But what if "foo" already exists when this step is run? The execution will go:
+ 1. Stairway runs `do()` and call create "foo". That results in a "foo already exists"
+ exception.
+ 1. Stairway catches the exception and decides the flight is failed. It
+ transitions to undoing.
+ 1. Stairway runs `undo()` and calls delete "foo". It deletes the pre-existing object.
```

Comment:
aside: I wonder if there are patterns we could use to bake this possibility out. That is, a DSL or something that understands and avoids the pitfalls.